Cannot `map` to a different type

Adding the following code to optional_monadic.cpp:

```
auto a = make();
auto f = [](auto x){return "oh no";};
scelta::map(a, f);
```

results in this error on GCC 7.2 (cut down for brevity):

```
In file included from /home/simon/scelta/include/scelta/utils.hpp:13:0,
                 from /home/simon/scelta/test/optional/../variant_test_utils.hpp:9,
                 from /home/simon/scelta/test/optional/optional_monadic.cpp:2:
/home/simon/scelta/include/scelta/./utils/optional_monadic.hpp:29:20: note: candidate: template<class Optional, class T, class F> constexpr decltype ((scelta::is_nullopt(o) ? forward<decltype (def)>(def) : forward<decltype (f)>(f)(scelta::impl::access_optional(forward<decltype (o)>(o))))) scelta::map_or(Optional&&, T&&, F&&)
     constexpr auto map_or(Optional&& o, T&& def, F&& f)
                    ^~~~~~
/home/simon/scelta/include/scelta/./utils/optional_monadic.hpp:29:20: note:   template argument deduction/substitution failed:
In file included from /home/simon/scelta/include/scelta/./support/./variant/./enable/../../../traits/adt/../dispatch.hpp:8:0,
                 from /home/simon/scelta/include/scelta/./support/./variant/./enable/../../../traits/adt/valid.hpp:7,
                 from /home/simon/scelta/include/scelta/./support/./variant/./enable/std.hpp:10,
                 from /home/simon/scelta/include/scelta/./support/./variant/std.hpp:13,
                 from /home/simon/scelta/include/scelta/./support/variant.hpp:7,
                 from /home/simon/scelta/include/scelta/support.hpp:7,
                 from /home/simon/scelta/test/optional/../variant_test_utils.hpp:8,
                 from /home/simon/scelta/test/optional/optional_monadic.cpp:2:
/home/simon/scelta/include/scelta/./utils/optional_monadic.hpp: In substitution of ‘template<class Optional, class T, class F> constexpr decltype ((scelta::is_nullopt(o) ? forward<decltype (def)>(def) : forward<decltype (f)>(f)(scelta::impl::access_optional(forward<decltype (o)>(o))))) scelta::map_or(Optional&&, T&&, F&&) [with Optional = boost::optional<int>&; T = boost::optional<int>; F = main()::<lambda(auto:15)> [with auto:15 = test::impl::maker_t<boost::optional<int> >]::<lambda(auto:16)>&]’:
/home/simon/scelta/include/scelta/./utils/optional_monadic.hpp:39:5:   required by substitution of ‘template<class Optional, class F> constexpr decltype (scelta::map_or(forward<decltype (o)>(o), std::decay_t<_Tp>{}, forward<decltype (f)>(f))) scelta::map(Optional&&, F&&) [with Optional = boost::optional<int>&; F = main()::<lambda(auto:15)> [with auto:15 = test::impl::maker_t<boost::optional<int> >]::<lambda(auto:16)>&]’
/home/simon/scelta/test/optional/optional_monadic.cpp:130:28:   required from ‘main()::<lambda(auto:15)> [with auto:15 = test::impl::maker_t<boost::optional<int> >]’
/home/simon/scelta/test/optional/../variant_test_utils.hpp:117:28:   required from ‘test::with_all_optional_implementations(TF&&)::<lambda(auto:14)> [with auto:14 = test::impl::unpack_alternatives<test::impl::alternatives_t<int> >::applier<boost::optional>; TAlternative = int; TF = main()::<lambda(auto:15)>]’
/home/simon/scelta/test/optional/../variant_test_utils.hpp:91:10:   required from ‘void test::instantiate_with_all_optional_implementations(TF&&) [with TestCase = test::impl::unpack_alternatives<test::impl::alternatives_t<int> >::applier; TF = test::with_all_optional_implementations(TF&&) [with TAlternative = int; TF = main()::<lambda(auto:15)>]::<lambda(auto:14)>]’
/home/simon/scelta/test/optional/../variant_test_utils.hpp:116:80:   required from ‘void test::with_all_optional_implementations(TF&&) [with TAlternative = int; TF = main()::<lambda(auto:15)>]’
/home/simon/scelta/test/optional/optional_monadic.cpp:134:10:   required from here
/home/simon/scelta/include/scelta/./utils/optional_monadic.hpp:31:23: error: operands to ?: have different types ‘boost::optional<int>’ and ‘const char*’
         is_nullopt(o) ? FWD(def)
         ~~~~~~~~~~~~~~^~~~~~~~~~
                       : FWD(f)(impl::access_optional(FWD(o)))

```

Expected behaviour: Compile and run

`and_then` has a similar issue.

A fix would be to get the return type of calling `f` with the value in the optional and explicitly construct the correct type on both branches of the conditional.

--- a/example/optional_cat.cpp
+++ b/example/optional_cat.cpp
@@ -17,51 +17,63 @@ bool success()
 template <typename T>
 using optional = example::optional<T>;
 
-struct image_view
+struct cymk_image_view
 {
 };
 
-optional<image_view> simulate_operation()
+struct rgb_image_view
 {
-    return success() ? optional<image_view>{image_view{}}
+};
+
+template <typename T>
+optional<T> simulate_operation()
+{
+    return success() ? optional<T>{T{}}
                      : scelta::nullopt;
 }
 
-optional<image_view> crop_to_cat(image_view)
+optional<rgb_image_view> to_rgb(cymk_image_view)
+{
+    std::cout << "converting...\n";
+    return simulate_operation<rgb_image_view>();
+}
+
+optional<rgb_image_view> crop_to_cat(rgb_image_view)
 {
     std::cout << "cropping...\n";
-    return simulate_operation();
+    return simulate_operation<rgb_image_view>();
 }
 
-optional<image_view> add_bow_tie(image_view)
+optional<rgb_image_view> add_bow_tie(rgb_image_view)
 {
     std::cout << "adding bow tie...\n";
-    return simulate_operation();
+    return simulate_operation<rgb_image_view>();
 }
 
-optional<image_view> make_eyes_sparkle(image_view)
+optional<rgb_image_view> make_eyes_sparkle(rgb_image_view)
 {
     std::cout << "making eyes sparkle...\n";
-    return simulate_operation();
+    return simulate_operation<rgb_image_view>();
 }
 
-image_view make_smaller(image_view)
+rgb_image_view make_smaller(rgb_image_view)
 {
     std::cout << "making smaller...\n";
     return {};
 }
 
-image_view add_rainbow(image_view)
+rgb_image_view add_rainbow(rgb_image_view)
 {
     std::cout << "adding rainbow...\n";
     return {};
 }
 
 // clang-format off
-optional<image_view> get_cute_cat(image_view img)
+optional<rgb_image_view> get_cute_cat(cymk_image_view img)
 {
     using namespace scelta::infix;
-    return crop_to_cat(img)
+    return to_rgb(img)
+         | and_then(crop_to_cat)
          | and_then(add_bow_tie)
          | and_then(make_eyes_sparkle)
          | map(make_smaller)

--- a/include/scelta/meta.hpp
+++ b/include/scelta/meta.hpp
@@ -9,4 +9,5 @@
 #include "./meta/fn_ref_wrapper.hpp"
 #include "./meta/forward_like.hpp"
 #include "./meta/replace_all.hpp"
+#include "./meta/type_wrapper.hpp"
 #include "./meta/y_combinator.hpp"

--- a/include/scelta/meta/type_wrapper.hpp
+++ b/include/scelta/meta/type_wrapper.hpp
@@ -1,0 +1,20 @@
+// Copyright (c) 2017 Vittorio Romeo
+// MIT License |  https://opensource.org/licenses/MIT
+// http://vittorioromeo.info | vittorio.romeo@outlook.com
+
+#pragma once
+
+namespace scelta::meta
+{
+    template <typename T>
+    struct type_wrapper
+    {
+        using type = T;
+    };
+
+    template <typename T>
+    inline constexpr type_wrapper<T> t{};
+
+    template <typename T>
+    using unwrap = typename T::type;
+}

--- a/include/scelta/nonrecursive/match.hpp
+++ b/include/scelta/nonrecursive/match.hpp
@@ -22,14 +22,14 @@ namespace scelta
             {
             }
 
-#define DEFINE_BOUND_VISITOR_CALL(m_refq)                                  \
-    template <typename... Variants>                                        \
-    constexpr auto operator()(Variants&&... variants) m_refq               \
-    SCELTA_NOEXCEPT_AND_TRT(                                               \
+#define DEFINE_BOUND_VISITOR_CALL(m_refq)                                                \
+    template <typename... Variants>                                                      \
+    constexpr auto operator()(Variants&&... variants) m_refq                             \
+    SCELTA_NOEXCEPT_AND_TRT(                                                             \
         ::scelta::nonrecursive::visit(std::declval<Visitor m_refq>(), FWD(variants)...)) \
-    {                                                                      \
+    {                                                                                    \
         return ::scelta::nonrecursive::visit(                                            \
-            static_cast<Visitor m_refq>(*this), FWD(variants)...);         \
+            static_cast<Visitor m_refq>(*this), FWD(variants)...);                       \
     }
 
             DEFINE_BOUND_VISITOR_CALL(&)

--- a/include/scelta/support/optional/enable/type_safe.hpp
+++ b/include/scelta/support/optional/enable/type_safe.hpp
@@ -47,5 +47,22 @@ namespace scelta::traits
                 )
             // clang-format on
         };
+
+        template <typename T>
+        struct make<::type_safe::basic_optional<T>>
+        {
+            // `T` in this context is a storage policy. In order to create a new
+            // optional instance, we can use `rebind` to change the stored type.
+
+            // clang-format off
+            template <typename X>
+            constexpr auto operator()(X&& x)
+                SCELTA_RETURNS(
+                    ::type_safe::basic_optional<
+                        typename T::template rebind<std::decay_t<X>>
+                    >{FWD(x)}
+                )
+            // clang-format on
+        };
     }
 }

--- a/include/scelta/traits/adt/visit_optional.hpp
+++ b/include/scelta/traits/adt/visit_optional.hpp
@@ -4,7 +4,7 @@
 
 #pragma once
 
-#include "../../utils/optional_utils.hpp"
+#include "../../utils/visit_optional.hpp"
 
 namespace scelta::traits::adt
 {

--- a/include/scelta/traits/optional.hpp
+++ b/include/scelta/traits/optional.hpp
@@ -4,23 +4,6 @@
 
 #pragma once
 
-#include "../utils/returns.hpp"
-#include "./dispatch.hpp"
-
-namespace scelta::traits::optional
-{
-    template <typename>
-    struct access
-    {
-        // clang-format off
-        template <typename Optional>
-        constexpr auto operator()(Optional&& o)
-            SCELTA_RETURNS(
-                *FWD(o)
-            )
-        // clang-format on
-    };
-
-    template <typename T>
-    inline constexpr traits::impl::dispatch<access<T>> access_v{};
-}
+#include "./optional/access.hpp"
+#include "./optional/make.hpp"
+#include "./optional/make_unset.hpp"

--- a/include/scelta/traits/optional/access.hpp
+++ b/include/scelta/traits/optional/access.hpp
@@ -1,0 +1,26 @@
+// Copyright (c) 2017 Vittorio Romeo
+// MIT License |  https://opensource.org/licenses/MIT
+// http://vittorioromeo.info | vittorio.romeo@outlook.com
+
+#pragma once
+
+#include "../../utils/returns.hpp"
+#include "../dispatch.hpp"
+
+namespace scelta::traits::optional
+{
+    template <typename>
+    struct access
+    {
+        // clang-format off
+        template <typename Optional>
+        constexpr auto operator()(Optional&& o)
+            SCELTA_RETURNS(
+                *FWD(o)
+            )
+        // clang-format on
+    };
+
+    template <typename T>
+    inline constexpr traits::impl::dispatch<access<T>> access_v{};
+}

--- a/include/scelta/traits/optional/make.hpp
+++ b/include/scelta/traits/optional/make.hpp
@@ -1,0 +1,30 @@
+// Copyright (c) 2017 Vittorio Romeo
+// MIT License |  https://opensource.org/licenses/MIT
+// http://vittorioromeo.info | vittorio.romeo@outlook.com
+
+#pragma once
+
+#include "../../utils/returns.hpp"
+#include "../dispatch.hpp"
+
+namespace scelta::traits::optional
+{
+    template <typename>
+    struct make;
+
+    template <template <typename...> typename Optional, typename T,
+        typename... Ts>
+    struct make<Optional<T, Ts...>>
+    {
+        // clang-format off
+        template <typename... Xs>
+        constexpr auto operator()(Xs&&... xs)
+            SCELTA_RETURNS(
+                Optional<std::decay_t<Xs>...>{FWD(xs)...}
+            )
+        // clang-format on
+    };
+
+    template <typename T>
+    inline constexpr traits::impl::dispatch<make<T>> make_v{};
+}

--- a/include/scelta/traits/optional/make_unset.hpp
+++ b/include/scelta/traits/optional/make_unset.hpp
@@ -1,0 +1,31 @@
+// Copyright (c) 2017 Vittorio Romeo
+// MIT License |  https://opensource.org/licenses/MIT
+// http://vittorioromeo.info | vittorio.romeo@outlook.com
+
+#pragma once
+
+#include "../../utils/returns.hpp"
+#include "../dispatch.hpp"
+#include "../../meta/type_wrapper.hpp"
+
+namespace scelta::traits::optional
+{
+    template <typename>
+    struct make_unset;
+
+    template <template <typename...> typename Optional, typename T,
+        typename... Ts>
+    struct make_unset<Optional<T, Ts...>>
+    {
+        // clang-format off
+        template <typename X>
+        constexpr auto operator()(X)
+            SCELTA_RETURNS(
+                Optional<scelta::meta::unwrap<X>>{}
+            )
+        // clang-format on
+    };
+
+    template <typename T>
+    inline constexpr traits::impl::dispatch<make_unset<T>> make_unset_v{};
+}

--- a/include/scelta/utils.hpp
+++ b/include/scelta/utils.hpp
@@ -15,3 +15,4 @@
 #include "./utils/optional_utils.hpp"
 #include "./utils/overload.hpp"
 #include "./utils/returns.hpp"
+#include "./utils/visit_optional.hpp"

--- a/include/scelta/utils/optional_monadic.hpp
+++ b/include/scelta/utils/optional_monadic.hpp
@@ -13,43 +13,91 @@
 #include "./optional_utils.hpp"
 #include "./returns.hpp"
 
+// TODO: remove?
+namespace scelta::detail
+{
+    template <typename Optional>
+    struct change_item_type;
+
+    template <template <typename...> typename Optional, typename T,
+        typename... Ts>
+    struct change_item_type<Optional<T, Ts...>>
+    {
+        template <typename U>
+        using apply = Optional<U, Ts...>;
+    };
+
+    template <typename Optional, typename T>
+    using change_item_type_t = typename change_item_type<Optional>::template apply<T>;
+
+    template <typename Optional>
+    using optional_item_type_t =
+        decltype(impl::access_optional(std::declval<Optional>()));
+
+    template <typename Optional, typename F>
+    using unwrap_optional_invoke_result_t =
+        std::invoke_result_t<F, optional_item_type_t<Optional>>;
+
+    template <typename Optional, typename F>
+    using map_result_t = change_item_type_t<std::decay_t<Optional>,
+        unwrap_optional_invoke_result_t<Optional, F>>;
+}
+
 namespace scelta
 {
     // clang-format off
     /// @brief Returns `f(*o)` if `o` is set, `f_def()` otherwise.
+    /// @details Signatures:
+    /// * `o: Optional<T>`
+    /// * `f_def: () -> U`
+    /// * `f: (T) -> U`
     template <typename Optional, typename FD, typename F>
     constexpr auto map_or_else(Optional&& o, FD&& f_def, F&& f)
     SCELTA_RETURNS(
-        is_nullopt(o) ? FWD(f_def)()
-                      : FWD(f)(impl::access_optional(FWD(o)))
+        is_nullopt(o) ? make_like<Optional>(FWD(f_def)())
+                      : make_like<Optional>(FWD(f)(impl::access_optional(FWD(o))))
     )
 
     /// @brief Returns `f(*o)` if `o` is set, `def` otherwise.
+    /// @details Signatures:
+    /// * `o: Optional<T>`
+    /// * `def: U`
+    /// * `f: (T) -> U`
     template <typename Optional, typename T, typename F>
     constexpr auto map_or(Optional&& o, T&& def, F&& f)
     SCELTA_RETURNS(
-        is_nullopt(o) ? FWD(def)
-                      : FWD(f)(impl::access_optional(FWD(o)))
+        is_nullopt(o) ? make_like<Optional>(FWD(def))
+                      : make_like<Optional>(FWD(f)(impl::access_optional(FWD(o))))
     )
 
     /// @brief Returns `Optional{f(*o)}` if `o` is set, an empty
-    // `Optional` otherwise.
+    /// optional otherwise.
+    /// @details Signatures:
+    /// * `o: Optional<T>`
+    /// * `f: (T) -> U`
     template <typename Optional, typename F>
     constexpr auto map(Optional&& o, F&& f)
     SCELTA_RETURNS(
-        map_or(FWD(o), std::decay_t<Optional>{}, FWD(f))
+         is_nullopt(o) ? make_unset_like<Optional, decltype(FWD(f)(impl::access_optional(FWD(o))))>()
+                       : make_like<Optional>(FWD(f)(impl::access_optional(FWD(o))))
     )
 
     /// @brief Returns `f(*o)` if `o` is set, an empty `Optional`
     /// otherwise.
+    /// @details Signatures:
+    /// * `o: Optional<T>`
+    /// * `f: (T) -> Optional<U>`
     template <typename Optional, typename F>
     constexpr auto and_then(Optional&& o, F&& f)
     SCELTA_RETURNS(
-        is_nullopt(o) ? std::decay_t<Optional>{}
+        is_nullopt(o) ? std::decay_t<decltype(FWD(f)(impl::access_optional(FWD(o))))>{}
                       : FWD(f)(impl::access_optional(FWD(o)))
     )
 
     /// @brief Returns `ob` if `o` is set, an empty `OptB` otherwise.
+    /// @details Signatures:
+    /// * `o: Optional<T>`
+    /// * `ob: Optional<U>`
     template <typename Optional, typename OptB>
     constexpr auto and_(Optional&& o, OptB&& ob)
     SCELTA_RETURNS(
@@ -57,6 +105,9 @@ namespace scelta
     )
 
     /// @brief Returns `o` if `o` is set, `f()` otherwise.
+    /// @details Signatures:
+    /// * `o: Optional<T>`
+    /// * `f: () -> Optional<U>`
     template <typename Optional, typename F>
     constexpr auto or_else(Optional&& o, F&& f)
     SCELTA_RETURNS(
@@ -64,6 +115,9 @@ namespace scelta
     )
 
     /// @brief Returns `o` if `o` is set, `def` otherwise.
+    /// @details Signatures:
+    /// * `o: Optional<T>`
+    /// * `def: Optional<U>`
     template <typename Optional, typename Def>
     constexpr auto or_(Optional&& o, Def&& def)
     SCELTA_RETURNS(

--- a/include/scelta/utils/optional_monadic_infix.hpp
+++ b/include/scelta/utils/optional_monadic_infix.hpp
@@ -13,17 +13,20 @@ namespace scelta::infix
         template <typename F>
         struct infix_f : F
         {
-            constexpr infix_f(F&& f) : F{FWD(f)}
+            constexpr infix_f(F&& f) noexcept(noexcept(F{FWD(f)}))
+                : F{FWD(f)}
             {
             }
         };
     }
 
+    // clang-format off
     template <typename Optional, typename F>
     constexpr auto operator|(Optional&& o, detail::infix_f<F>&& f)
-    {
-        return f(FWD(o));
-    }
+        SCELTA_RETURNS(
+           f(FWD(o))
+        )
+    // clang-format on
 
 #define SCELTA_DEFINE_INFIX(name)                                              \
     template <typename... Ts>                                                  \

--- a/include/scelta/utils/optional_utils.hpp
+++ b/include/scelta/utils/optional_utils.hpp
@@ -4,58 +4,28 @@
 
 #pragma once
 
-#include "../meta/forward_like.hpp"
-#include "../traits/adt/is_visitable.hpp"
-#include "../traits/adt/valid.hpp"
-#include "./access_optional.hpp"
-#include "./assert.hpp"
+#include "../meta/type_wrapper.hpp"
+#include "../traits/optional.hpp"
 #include "./returns.hpp"
-#include "./nullopt.hpp"
+#include <type_traits>
 
-namespace scelta::impl
+namespace scelta
 {
-    // Invokes the continuation `c` with the result of `f(<"unpacked" o>)`.
-    template <typename F, typename Continuation, typename Optional>
-    constexpr auto call_with_optional(F&& f, Continuation&& c, Optional&& o)
-        SCELTA_RETURNS(
-            o ? FWD(c)(FWD(f)(meta::forward_like<Optional>(::scelta::impl::access_optional(o))))
-              : FWD(c)(FWD(f)(nullopt))
-        )
-
     // clang-format off
-    template <typename Visitor>
-    constexpr auto visit_optional(Visitor&& visitor)
+    template <typename Optional, typename... Ts>
+    constexpr auto make_like(Ts&&... xs)
         SCELTA_RETURNS(
-            FWD(visitor)()
+            ::scelta::traits::optional::make_v<std::decay_t<Optional>>(
+                FWD(xs)...
+            )
         )
 
-    // TODO: noexcept incorrect
-    template <typename Visitor, typename Optional, typename... Optionals>
-    constexpr auto visit_optional(
-        Visitor&& visitor, Optional&& o, Optionals&&... os) ->decltype(auto)
-        // SCELTA_NOEXCEPT_AND_TRT(FWD(visitor)(to_nullopt(), to_nullopt(os)...))
-    {
-        return
-            SCELTA_CONSTEXPR_ASSERT(
-                ::scelta::traits::adt::valid_v<std::decay_t<Optional>>(o, os...)
-            ),
-            call_with_optional(
-            // Binds the first "unpacked" optional to the visitor.
-            // Returns a the next visitor with the bound value, which will
-            // be passed to the continuation function.
-            [&](auto&& x) noexcept
-            {
-                return [&](auto&&... xs) SCELTA_RETURNS(
-                    FWD(visitor)(FWD(x), FWD(xs)...)
-                );
-            },
-
-            // Continuation: recurse upon the newly bound visitor.
-            [&](auto&& bound_visitor) SCELTA_RETURNS(
-                visit_optional(FWD(bound_visitor), FWD(os)...)
-            ),
-            FWD(o)
-        );
-    }
+    template <typename Optional, typename T>
+    constexpr auto make_unset_like()
+        SCELTA_RETURNS(
+            ::scelta::traits::optional::make_unset_v<std::decay_t<Optional>>(
+                ::scelta::meta::t<T>
+            )
+        )
     // clang-format on
 }

--- a/include/scelta/utils/visit_optional.hpp
+++ b/include/scelta/utils/visit_optional.hpp
@@ -1,0 +1,61 @@
+// Copyright (c) 2017 Vittorio Romeo
+// MIT License |  https://opensource.org/licenses/MIT
+// http://vittorioromeo.info | vittorio.romeo@outlook.com
+
+#pragma once
+
+#include "../meta/forward_like.hpp"
+#include "../traits/adt/is_visitable.hpp"
+#include "../traits/adt/valid.hpp"
+#include "./access_optional.hpp"
+#include "./assert.hpp"
+#include "./nullopt.hpp"
+#include "./returns.hpp"
+
+namespace scelta::impl
+{
+    // clang-format off
+    // Invokes the continuation `c` with the result of `f(<"unpacked" o>)`.
+    template <typename F, typename Continuation, typename Optional>
+    constexpr auto call_with_optional(F&& f, Continuation&& c, Optional&& o)
+        SCELTA_RETURNS(
+            o ? FWD(c)(FWD(f)(meta::forward_like<Optional>(::scelta::impl::access_optional(o))))
+              : FWD(c)(FWD(f)(nullopt))
+        )
+
+    template <typename Visitor>
+    constexpr auto visit_optional(Visitor&& visitor)
+        SCELTA_RETURNS(
+            FWD(visitor)()
+        )
+
+    // TODO: noexcept incorrect
+    template <typename Visitor, typename Optional, typename... Optionals>
+    constexpr auto visit_optional(
+        Visitor&& visitor, Optional&& o, Optionals&&... os) ->decltype(auto)
+        // SCELTA_NOEXCEPT_AND_TRT(FWD(visitor)(to_nullopt(), to_nullopt(os)...))
+    {
+        return
+            SCELTA_CONSTEXPR_ASSERT(
+                ::scelta::traits::adt::valid_v<std::decay_t<Optional>>(o, os...)
+            ),
+            call_with_optional(
+            // Binds the first "unpacked" optional to the visitor.
+            // Returns a the next visitor with the bound value, which will
+            // be passed to the continuation function.
+            [&](auto&& x) noexcept
+            {
+                return [&](auto&&... xs) SCELTA_RETURNS(
+                    FWD(visitor)(FWD(x), FWD(xs)...)
+                );
+            },
+
+            // Continuation: recurse upon the newly bound visitor.
+            [&](auto&& bound_visitor) SCELTA_RETURNS(
+                visit_optional(FWD(bound_visitor), FWD(os)...)
+            ),
+            FWD(o)
+        );
+    }
+    // clang-format on
+}

--- a/test/optional/optional_monadic.cpp
+++ b/test/optional/optional_monadic.cpp
@@ -13,6 +13,11 @@ TEST_MAIN()
     with_all_optional_implementations<int>( //
         [](auto make)                       //
         {
+            auto make_other = [](auto x)
+            {
+                return ::scelta::make_like<decltype(make())>(x);
+            };
+
             // map_or_else, set
             {
                 auto f_def = [&]{ return 42; };
@@ -31,6 +36,15 @@ TEST_MAIN()
                 EXPECT_TRUE((o | scelta::infix::map_or_else(f_def, f)) == 42);
             }
 
+            // map_or_else with a different type
+            {
+                auto f_def = [&]{ return 'b'; };
+                auto f = [](int){ return 'a'; };
+                auto o = make(0);
+                EXPECT_TRUE(scelta::map_or_else(o, f_def, f) == make_other('a'));
+                EXPECT_TRUE((o | scelta::infix::map_or_else(f_def, f)) == make_other('a'));
+            }
+
             // map_or, set
             {
                 auto f = [](int x){ return x + 1; };
@@ -45,6 +59,14 @@ TEST_MAIN()
                 auto o = make();
                 EXPECT_TRUE(scelta::map_or(o, 42, f) == 42);
                 EXPECT_TRUE((o | scelta::infix::map_or(42, f)) == 42);
+            }
+
+            // map_or with a different type
+            {
+                auto f = [](int){ return 'a'; };
+                auto o = make(0);
+                EXPECT_TRUE(scelta::map_or(o, 'x', f) == make_other('a'));
+                EXPECT_TRUE((o | scelta::infix::map_or('x', f)) == make_other('a'));
             }
 
             // map, set
@@ -63,6 +85,13 @@ TEST_MAIN()
                 EXPECT_TRUE((o | scelta::infix::map(f)) == make());
             }
 
+            // map to a different type
+            {
+                auto o = make(0);
+                auto f = [](int){ return 'a'; };
+                EXPECT_TRUE(scelta::map(o, f) == make_other('a'));
+            }
+
             // and_then, set
             {
                 auto f = [&](int x){ return make(x + 1); };
@@ -79,20 +108,28 @@ TEST_MAIN()
                 EXPECT_TRUE((o | scelta::infix::and_then(f)) == make());
             }
 
+            // and_then with a different type
+            {
+                auto f = [&](int){ return make_other('a'); };
+                auto o = make(0);
+                EXPECT_TRUE(scelta::and_then(o, f) == make_other('a'));
+                EXPECT_TRUE((o | scelta::infix::and_then(f)) == make_other('a'));
+            }
+
             // and_, set
             {
                 auto a = make(10);
-                auto b = make('a');
+                auto b = make_other('a');
                 EXPECT_TRUE(scelta::and_(a, b) == b);
-                EXPECT_TRUE((a | scelta::infix::and_(b)) == make(b));
+                EXPECT_TRUE((a | scelta::infix::and_(b)) == make_other('a'));
             }
 
             // and_, unset
             {
                 auto a = make();
-                auto b = make('a');
-                EXPECT_TRUE(scelta::and_(a, b) == make());
-                EXPECT_TRUE((a | scelta::infix::and_(b)) == make());
+                auto b = make_other('a');
+                EXPECT_TRUE(scelta::and_(a, b) == decltype(b){});
+                EXPECT_TRUE((a | scelta::infix::and_(b)) == decltype(b){});
             }
 
             // or_else, set

--- a/test/variant_test_utils.hpp
+++ b/test/variant_test_utils.hpp
@@ -91,9 +91,9 @@ namespace test
         f(TestCase<boost::optional>{});
 #endif
 
-#if defined(SCELTA_SUPPORT_OPTIONAL_TYPE_SAFE)
-        f(TestCase<type_safe::optional>{});
-#endif
+// #if defined(SCELTA_SUPPORT_OPTIONAL_TYPE_SAFE)
+//         f(TestCase<type_safe::optional>{});
+// #endif
 
 #if defined(SCELTA_SUPPORT_OPTIONAL_TL)
         f(TestCase<tl::optional>{});

